applications/serial_lte_modem/src/slm_at_tcp_proxy.c

```diff
@@ -435,123 +443,255 @@ static int tcp_data_save(uint8_t *data, uint32_t length)
 				peer_addr);
 			rsp_send(rsp_buf, strlen(rsp_buf));
 		}
 		proxy.sock_peer = ret;
+		LOG_DBG("New connection - %d",
+			proxy.sock_peer);
+		fds[nfds].fd = proxy.sock_peer;
+		fds[nfds].events = POLLIN;
+		nfds++;
 		/* Start a one-shot timer to close the connection */
-		k_timer_start(&conn_timer, K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+		k_timer_start(&conn_timer,
+			      K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
 			      K_NO_WAIT);
-	}
-
-	if (proxy.role == AT_TCP_ROLE_SERVER) {
-		sock = proxy.sock_peer;
 	} else {
-		sock = proxy.sock;
+		char data[NET_IPV4_MTU];
+
+		k_timer_stop(&conn_timer);
+		ret = recv(fds[infd].fd, data, NET_IPV4_MTU, 0);
+		if (ret < 0) {
+			LOG_WRN("recv() error: %d", -errno);
+			return -errno;
+		}
+		if (ret == 0) {
+			goto recv_done;
+		}
+		tcp_data_handle(data, ret);
+recv_done:
+		/* Restart conn timer */
+		LOG_ERR("restart timer: POLLIN");
+		k_timer_start(&conn_timer,
+			      K_SECONDS(CONFIG_SLM_TCP_CONN_TIME),
+			      K_NO_WAIT);
 	}
-	fds.fd = sock;
-	fds.events = POLLIN;
+
+	return ret;
+}
+
+/* TCP server thread */
+static void tcpsvr_thread_func(void *p1, void *p2, void *p3)
+{
+	int ret, current_size;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	fds[0].fd = proxy.sock;
+	fds[0].events = POLLIN;
+	nfds = 1;
 	ring_buf_reset(&data_buf);
 	while (true) {
 		if (proxy.role == AT_TCP_ROLE_SERVER &&
 			k_timer_status_get(&conn_timer) > 0) {
 			k_timer_stop(&conn_timer);
 			LOG_INF("Connecion timeout");
-			sprintf(rsp_buf, "#XTCPSVR: timeout\r\n");
+			sprintf(rsp_buf, "#XTCPSVR: %d timeout\r\n", -ETIME);
 			rsp_send(rsp_buf, strlen(rsp_buf));
 			close(proxy.sock_peer);
-			goto thread_entry;
+			proxy.sock_peer = INVALID_SOCKET;
+			nfds--;
 		}
-		ret = poll(&fds, 1, MSEC_PER_SEC * CONFIG_SLM_TCP_POLL_TIME);
+		ret = poll(fds, nfds, MSEC_PER_SEC * CONFIG_SLM_TCP_POLL_TIME);
 		if (ret < 0) {  /* IO error */
-			LOG_WRN("poll() error: %d", ret);
+			LOG_WRN("poll() error: %d", -errno);
+			ret = -EIO;
+			goto exit;
+		}
+		if (ret == 0) {  /* timeout */
+			LOG_DBG("poll() timeout");
 			continue;
 		}
+		current_size = nfds;
+		for (int i = 0; i < current_size; i++) {
+			LOG_DBG("Poll events 0x%08x", fds[i].revents);
+			if ((fds[i].revents & POLLERR) == POLLERR) {
+				LOG_ERR("POLLERR: %d", i);
+				ret = -EIO;
+				goto exit;
+			}
+			if ((fds[i].revents & POLLHUP) == POLLHUP) {
+				LOG_DBG("Peer disconnect:%d", fds[i].fd);
+				sprintf(rsp_buf, "#XTCPSVR: %d"
+					" disconnected\r\n", fds[i].fd);
+				rsp_send(rsp_buf, strlen(rsp_buf));
+				close(fds[i].fd);
+				if (fds[i].fd != proxy.sock) {
+					k_timer_stop(&conn_timer);
+				}
+				nfds--;
+				continue;
+			}
+			if ((fds[i].revents & POLLNVAL) == POLLNVAL) {
+				nfds--;
+			}
+			if ((fds[i].revents & POLLIN) == POLLIN) {
+				ret = tcpsvr_input(i);
+				if (ret < 0) {
+					LOG_WRN("tcpsvr_input error: %d", ret);
+				}
+			}
+		}
+		if (nfds == 0) {
+			goto exit;
```

Comment:
There is an assumption here, when fds[0].sock is closed, fds[1].sock could have not be closed yet, but we still need to goto exit.

Reply:
Removed the logic to check nfds.